crimsobot/cogs/cringo.py

```diff
@@ -1,225 +1,423 @@
+                if player.mismatch_count >= 8:
+                    left_game_embed = self.remove_from_game(player.user.id)
+                    await ctx.send(embed=left_game_embed)
+                    continue
+
+                # tell players what happens at the end of this turn
+                try:
+                    await player.user.send(embed=turn_embed)
+                except discord.errors.Forbidden:
+                    self.remove_from_game(player.user.id)
 
         # final score + awards time!
         # nerf calculated such that division by zero never attained within player limit
-        x = len(list_of_players)
+        x = len(self.players)
         nerf = 0.05*x**2 - 2.05*x + 52  # (points / nerf = coin)
 
         # check if crimsoBOT home server
         if ctx.guild and ctx.guild.id == 552650672965943296:
             nerf = (2 - SERVER_BONUS) * nerf
 
-        score_string, winner = await cringo.cringo_scoreboard(list_of_players)
-
-        embed = c.crimbed(
-            title='**{}CRINGO!** FINAL SCORE'.format(CRINGO_RULES['name'][card_size]),
-            descr=score_string,
-            footer=f'Your points/{nerf:.1f}=your crimsoCOIN winnings!',
-            thumb_name=CRINGO_RULES['thumb'][card_size],
-            color_name=CRINGO_RULES['color'][card_size],
-        )
+        # process all scores & winnings
+        for player in self.players:
+            player.winnings = 0 if self.cursed else player.score / nerf
+            await crimsogames.win(player.user, player.winnings)
+            self.cog.all_players.discard(player.user.id)  # some final cleanup
 
-        # for some reason, a for loop wasn't doing the trick here...
-        while len(list_of_players) != 0:
-            # pull a player...
-            player = list_of_players[0]
-            # ...trip this is they are the winner...
-            win = player.user == winner
-            # ...calcuilate and give out winnings
-            winning_amount = player.score/nerf
-            await crimsogames.win(player.user, winning_amount)
-            # ...do stats but only if regular cringo...
-            if card_size == 4:
-                await cringo.cringo_stats(player, winning_amount, win)
-            # ...and finally remove them from list
-            await cringo.player_remove(list_of_players, player)
+        # now that we have our winnings calculated, generate a scoreboard with winnings and run through players again
+        # to record stats for them
+        final_scoreboard = self.get_scoreboard(game_finished=True)
+        if self.card_size == 4:
+            for player in self.players:
+                await cringo.cringo_stats(player, player == final_scoreboard.winner)
 
+        embed = self.generate_game_over_embed(final_scoreboard)
         await ctx.send(embed=embed)
 
+
+class Cringo(commands.Cog):
+    def __init__(self, bot: CrimsoBOT):
+        self.bot = bot
+        self.all_players: Set[CringoPlayer] = set()
```

Comment:
I may be wrong, but it looks like this can move into the `CringoGame` class? Then we don't have to rely on the Cog in CrimsoGame. Also it means we're not using the dumb `get_cog` service locator.